{{cookiecutter.project_name}}/{{cookiecutter.package_name}}/__init__.py

```diff
@@ -3,93 +3,117 @@
-            'metadata' : [],
-            'key_properties': []
-        }
-        streams.append(catalog_entry)
-
-    return {'streams': streams}
-
-def get_selected_streams(catalog):
-    '''
-    Gets selected streams.  Checks schema's 'selected' first (legacy)
-    and then checks metadata (current), looking for an empty breadcrumb
-    and mdata with a 'selected' entry
-    '''
-    selected_streams = []
-    for stream in catalog.streams:
-        stream_metadata = metadata.to_map(stream.metadata)
-        # stream metadata will have an empty breadcrumb
-        if metadata.get(stream_metadata, (), "selected"):
-            selected_streams.append(stream.tap_stream_id)
+        key_properties = []
+        streams.append(
+            CatalogEntry(
+                tap_stream_id=schema_name,
+                stream=schema_name,
+                schema=schema,
+                key_properties=key_properties,
+                metadata=stream_metadata,
+                replication_key=None,
+                is_view=None,
+                database=None,
+                table=None,
+                row_count=None,
+                stream_alias=None,
+                replication_method=None,
+            )
+        )
+    return Catalog(streams)
+
+
+def read_all_data():
+    """ Read all rows from upstream source and return data in proper json format """
+    # TODO: Replace with real logic to retrieve data
+    for dummy_value in range(10000):
+        yield {"id": dummy_value, "name": str(dummy_value)}
+
+
+def get_row_batches(batch_size=1):
+    """ Return lists of data rows according to batch_size """
+    queued = []
+    queued_count = 0
+    for item in read_all_data():
+        queued.append(item)
+        queued_count += 1
+        if queued_count >= batch_size:
+            yield queued
+            queued = []
+            queued_count = 0
+    if queued:
+        yield queued
 
-    return selected_streams
 
 def sync(config, state, catalog):
-
-    selected_stream_ids = get_selected_streams(catalog)
-
+    """ Sync data from tap source """
+    selected_stream_ids = catalog.get_selected_streams(state)
     # Loop over streams in catalog
     for stream in catalog.streams:
         stream_id = stream.tap_stream_id
-        stream_schema = stream.schema
-        if stream_id in selected_stream_ids:
-            # TODO: sync code for stream goes here...
-            LOGGER.info('Syncing stream:' + stream_id)
+        LOGGER.info("Syncing stream:" + stream_id)
+
+        # TODO: initialize key and bookmark columns
+        key_columns = ["id"]
+        bookmark_column = "id"
+
+        singer.write_schema(
+            stream_name=stream_id, schema=stream.schema, key_properties=key_columns,
+        )
+        for rows in get_row_batches(DEFAULT_BATCH_SIZE):
+            singer.write_records(stream_id, rows)
+            if bookmark_column:
+                singer.write_state({stream_id: row[bookmark_column]})
```

Comment:
`row` doesn't exist here. The safest option (due to varying API behavior) is to write state once, after all records for the stream have been written, so something like this would make sense (if the rows are sorted by `bookmark_column`):

```
        for rows in get_row_batches(DEFAULT_BATCH_SIZE):
            singer.write_records(stream_id, rows)
        if bookmark_column:
            singer.write_state({stream_id: rows[-1][bookmark_column]})
```

Reply:
Oh, I see. Yes, `rows` is available in my above and `rows` is not. However, in a proposed update, I'll be reverting this again to the singular `row`. When one cannot ensure the rows are already sorted by that key, is the best practice to write the stage with each row?

Reply:
When sorting by `bookmark_column` cannot be confirmed, writing the state at the end is the best practice.

Writing state with each row has the possibility of making the value jump around mid-sync, which could cause issues if the job is interrupted (e.g., machine goes down, power surge, network issues, exception in data retrieval code, etc.)

Reply:
I've updated the sample per this feedback. https://github.com/singer-io/singer-tap-template/pull/11/files#diff-1d2b794daaea7fec68ece66882499deaR76-R90